File: py_solutions_1-10/Euler_4.py
# Find the largest palindrome number made from the product of two 3-digit 
# numbers?

from __future__ import print_function

import timeit

try:
    range = xrange
except NameError:
    pass

start = timeit.default_timer()

def euler_4():
    high = 0
    for x in range(100, 1000):
        for y in range(x, 1000):
            product = x * y
            if product < high: continue
            if str(product) == str(product)[::-1]:
                high = product
    return high

print("Answer: {}".format(euler_4()))
stop = timeit.default_timer()
print("Time: {0:9.5f}".format(stop - start))

File: Java/java_solutions_1-10/Euler_4.java
/*
/   Project Euler problem #4 Largest palindrome product
/
/   Find the largest palindrome made from the product of two 3-digit numbers.
/
*/

public class Euler_4
{
    public static boolean isPalindrome(String candidate) {
        return new StringBuilder(candidate).reverse()
                                           .toString()
                                           .equals(candidate);
    }

    public static void main(String[] args)
    {
        long start = System.nanoTime();
        int largestPalindrome = 0;

        for (int i = 100; i < 1000; i++) { 
            for (int j = 100; j < i; j++) {
                int current = i * j;
                if (current > largestPalindrome && 
                    isPalindrome(String.valueOf(current))) { 
                    largestPalindrome = current;
                }
            }
        }
        long stop = System.nanoTime();
        System.out.println("Answer: " + largestPalindrome);
        System.out.printf("Time: %.4f\n", ((float) stop - start) / 1_000_000_000);
    }
}
